File: step_13_binary_trees/13.3_hard/P008_requirements_for_unique_bt/Solution.java
/**
 * Problem: Requirements for Unique Binary Tree
 * Difficulty: MEDIUM | XP: 25
 *
 * Given two types of traversals, determine whether they uniquely identify
 * a binary tree structure.
 *
 * Rules:
 *   - Preorder + Inorder  -> UNIQUE (classic reconstruct)
 *   - Postorder + Inorder -> UNIQUE
 *   - Preorder + Postorder -> NOT unique (ambiguous for non-full trees)
 *   - Any single traversal -> NOT unique
 *   - Level order + Inorder -> UNIQUE
 *   - Preorder + Postorder WITH full BT constraint -> UNIQUE
 *
 * This is a knowledge-based problem; the "approaches" demonstrate
 * reconstruction from different traversal pairs.
 *
 * @author DSA_Nova
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int v) { val = v; }
    }

    // ============================================================
    // APPROACH 1: BRUTE FORCE - Check uniqueness by enumeration (small trees)
    // Time: O(N!)  |  Space: O(N)
    // Enumerate all possible BST/BT structures and check traversal match.
    // NOTE: This is illustrative — practical answer is the rule-based approach below.
    // In practice: implement reconstruction and check if result is unique.
    // ============================================================
    public static String bruteForce(String t1, String t2) {
        // Rule-based: given traversal type names, return if combination is unique
        Set<String> unique = new HashSet<>(Arrays.asList(
            "preorder+inorder", "inorder+preorder",
            "postorder+inorder", "inorder+postorder",
            "levelorder+inorder", "inorder+levelorder"
        ));
        String key = (t1 + "+" + t2).toLowerCase();
        return unique.contains(key) ? "UNIQUE" : "NOT UNIQUE";
    }

    // ============================================================
    // APPROACH 2: OPTIMAL - Reconstruct from preorder + inorder
    // Time: O(N)  |  Space: O(N) with index map
    // Classic divide & conquer: root = preorder[0],
    // split inorder at root index to get left/right subtrees
    // ============================================================
    public static TreeNode buildFromPreIn(int[] preorder, int[] inorder) {
        Map<Integer, Integer> inIdx = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) inIdx.put(inorder[i], i);
        return buildPreIn(preorder, 0, preorder.length - 1,
                          0, inorder.length - 1, inIdx);
    }

    private static TreeNode buildPreIn(int[] pre, int preL, int preR,
                                        int inL, int inR, Map<Integer, Integer> inIdx) {
        if (preL > preR) return null;
        TreeNode root = new TreeNode(pre[preL]);
        int k = inIdx.get(root.val) - inL; // left subtree size
        root.left  = buildPreIn(pre, preL + 1,     preL + k, inL,     inL + k - 1, inIdx);
        root.right = buildPreIn(pre, preL + k + 1, preR,     inL + k + 1, inR,     inIdx);
        return root;
    }

    // ============================================================
    // APPROACH 3: BEST - Reconstruct from postorder + inorder
    // Time: O(N)  |  Space: O(N)
    // Root = postorder[last]; split inorder at root for left/right.
    // ============================================================
    public static TreeNode buildFromPostIn(int[] postorder, int[] inorder) {
        Map<Integer, Integer> inIdx = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) inIdx.put(inorder[i], i);
        return buildPostIn(postorder, 0, postorder.length - 1,
                           0, inorder.length - 1, inIdx);
    }

    private static TreeNode buildPostIn(int[] post, int postL, int postR,
                                         int inL, int inR, Map<Integer, Integer> inIdx) {
        if (postL > postR) return null;
        TreeNode root = new TreeNode(post[postR]);
        int k = inIdx.get(root.val) - inL;
        root.left  = buildPostIn(post, postL,     postL + k - 1, inL,         inL + k - 1, inIdx);
        root.right = buildPostIn(post, postL + k, postR - 1,     inL + k + 1, inR,         inIdx);
        return root;
    }

    // Helper: inorder traversal for verification
    static List<Integer> inorder(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        inorderHelper(root, res);
        return res;
    }
    static void inorderHelper(TreeNode n, List<Integer> r) {
        if (n == null) return;
        inorderHelper(n.left, r); r.add(n.val); inorderHelper(n.right, r);
    }

    public static void main(String[] args) {
        System.out.println("=== Requirements for Unique Binary Tree ===");

        // Rule-based uniqueness check
        System.out.println("\n-- Uniqueness Rules --");
        String[][] pairs = {
            {"preorder", "inorder"},
            {"postorder", "inorder"},
            {"preorder", "postorder"},
            {"levelorder", "inorder"},
            {"preorder", "preorder"},
        };
        for (String[] p : pairs) {
            System.out.println(p[0] + " + " + p[1] + " -> " + bruteForce(p[0], p[1]));
        }

        // Reconstruct tree from preorder + inorder
        int[] preorder = {3, 9, 20, 15, 7};
        int[] inorder  = {9, 3, 15, 20, 7};
        System.out.println("\n-- Reconstruct from preorder + inorder --");
        System.out.println("Preorder: " + Arrays.toString(preorder));
        System.out.println("Inorder:  " + Arrays.toString(inorder));
        TreeNode tree1 = buildFromPreIn(preorder, inorder);
        System.out.println("Inorder of reconstructed: " + inorder(tree1)); // [9,3,15,20,7]

        // Reconstruct from postorder + inorder
        int[] postorder = {9, 15, 7, 20, 3};
        System.out.println("\n-- Reconstruct from postorder + inorder --");
        System.out.println("Postorder: " + Arrays.toString(postorder));
        System.out.println("Inorder:   " + Arrays.toString(inorder));
        TreeNode tree2 = buildFromPostIn(postorder, inorder);
        System.out.println("Inorder of reconstructed: " + inorder(tree2)); // [9,3,15,20,7]
    }
}

File: step_13_binary_trees/13.3_hard/P008_requirements_for_unique_bt/solution.py
"""
Problem: Requirements for Unique Binary Tree
Difficulty: MEDIUM | XP: 25

Determine which pairs of traversals uniquely identify a binary tree.

Rules:
  UNIQUE:     Preorder + Inorder, Postorder + Inorder, LevelOrder + Inorder
  NOT UNIQUE: Preorder + Postorder (ambiguous for non-full trees)
  NOT UNIQUE: Any single traversal alone

The three approaches implement:
  1. Rule-based uniqueness checker
  2. Reconstruct from preorder + inorder
  3. Reconstruct from postorder + inorder
"""
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ============================================================
# APPROACH 1: BRUTE FORCE - Rule-based uniqueness determination
# Time: O(1)  |  Space: O(1)
# Knowledge-based: state which traversal pairs uniquely define a tree
# ============================================================
def brute_force(t1: str, t2: str) -> str:
    unique_pairs = {
        frozenset(["preorder", "inorder"]),
        frozenset(["postorder", "inorder"]),
        frozenset(["levelorder", "inorder"]),
    }
    pair = frozenset([t1.lower(), t2.lower()])
    return "UNIQUE" if pair in unique_pairs else "NOT UNIQUE"


# ============================================================
# APPROACH 2: OPTIMAL - Reconstruct tree from preorder + inorder
# Time: O(N)  |  Space: O(N) for index map
# Root = preorder[0]; find in inorder to split left/right subtrees.
# ============================================================
def optimal(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    in_idx = {v: i for i, v in enumerate(inorder)}

    def build(pre_l: int, pre_r: int, in_l: int, in_r: int) -> Optional[TreeNode]:
        if pre_l > pre_r:
            return None
        root_val = preorder[pre_l]
        root = TreeNode(root_val)
        k = in_idx[root_val] - in_l  # left subtree size
        root.left  = build(pre_l + 1,     pre_l + k, in_l,         in_l + k - 1)
        root.right = build(pre_l + k + 1, pre_r,     in_l + k + 1, in_r)
        return root

    return build(0, len(preorder) - 1, 0, len(inorder) - 1)


# ============================================================
# APPROACH 3: BEST - Reconstruct tree from postorder + inorder
# Time: O(N)  |  Space: O(N)
# Root = postorder[-1]; find in inorder to split left/right subtrees.
# ============================================================
def best(postorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    in_idx = {v: i for i, v in enumerate(inorder)}

    def build(post_l: int, post_r: int, in_l: int, in_r: int) -> Optional[TreeNode]:
        if post_l > post_r:
            return None
        root_val = postorder[post_r]
        root = TreeNode(root_val)
        k = in_idx[root_val] - in_l  # left subtree size
        root.left  = build(post_l,     post_l + k - 1, in_l,         in_l + k - 1)
        root.right = build(post_l + k, post_r - 1,     in_l + k + 1, in_r)
        return root

    return build(0, len(postorder) - 1, 0, len(inorder) - 1)


def inorder_traversal(root: Optional[TreeNode]) -> List[int]:
    res = []
    def dfs(node):
        if not node: return
        dfs(node.left); res.append(node.val); dfs(node.right)
    dfs(root)
    return res


if __name__ == "__main__":
    print("=== Requirements for Unique Binary Tree ===")

    print("\n-- Uniqueness Rules --")
    pairs = [
        ("preorder", "inorder"),
        ("postorder", "inorder"),
        ("preorder", "postorder"),
        ("levelorder", "inorder"),
        ("preorder", "preorder"),
    ]
    for t1, t2 in pairs:
        print(f"  {t1} + {t2} -> {brute_force(t1, t2)}")

    # Reconstruct from preorder + inorder
    preorder = [3, 9, 20, 15, 7]
    inorder  = [9, 3, 15, 20, 7]
    print("\n-- Reconstruct from preorder + inorder --")
    print(f"Preorder: {preorder}")
    print(f"Inorder:  {inorder}")
    tree = optimal(preorder, inorder)
    print(f"Inorder of result: {inorder_traversal(tree)}")   # [9,3,15,20,7]

    # Reconstruct from postorder + inorder
    postorder = [9, 15, 7, 20, 3]
    print("\n-- Reconstruct from postorder + inorder --")
    print(f"Postorder: {postorder}")
    print(f"Inorder:   {inorder}")
    tree = best(postorder, inorder)
    print(f"Inorder of result: {inorder_traversal(tree)}")   # [9,3,15,20,7]
